### DISCUSSION TITLE
Steady State Run Termination based on User-Defined Criterion Question

### DISCUSSION THREAD

**Discussion by wrkendrick** [idaholab/moose#17705] [Q&A General] [answered]
Hi all,

I've recently been working with a function that generates a temperature profile when given a temperature flux, and have used this in MOOSE to adjust boundary conditions every non-linear step until convergence is met. The tradition "convergence" however is the non-linear residual, which in this case takes too long to converge (I think because there's a natural oscillation of the direction in which the temperatures at the boundary change, making the residual explode in value even when the temperatures themselves are settling within +/- 0.1 degrees K).

I decided to follow in the footsteps of the Terminator.C user object and focus on using the change in boundary temperatures as my criterion for convergence. My version of the Terminator needs to look at a vector postprocessor (as it's multiple temperatures across the surface) so I stripped a lot of the smart functionality like the parser. It essentially works the same though, where if the criterion is satisfied then it should run `_fe_problem.terminateSolve()`. 

My problem is that this isn't working right now. It's running the above function but it doesn't interrupt the run. I think this is because the above is meant to kick in at the end of a time step, not in the middle of a solve. I could be wrong, but if I am correct is there any other function/structure I can use to tell MOOSE to cleanly wrap up the run? I can force it by making convergence act as an error, but I don't think it will output results in this case, so there's no real point.

Below is my vector-adjusted Terminator:

```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#include "libmesh/libmesh_config.h"

#ifdef LIBMESH_HAVE_FPARSER

#include "Terminator.h"
#include "MooseApp.h"
#include "MooseEnum.h"
#include "Executioner.h"

registerMooseObject("MooseApp", Terminator);

defineLegacyParams(Terminator);

InputParameters
Terminator::validParams()
{
  InputParameters params = GeneralUserObject::validParams();
  params.addClassDescription("Adjusted Terminator for HTPIPE purposes.");
  params.addRequiredParam<std::vector<VectorPostprocessorName>>(
      "vpp_names", "The list of name of VectorPostProcessesors to watch for convergence");
  params.addRequiredParam<double>("criterion", "Convergence criterion");
  MooseEnum failModeOption("HARD SOFT", "HARD");
  params.addParam<MooseEnum>(
      "fail_mode",
      failModeOption,
      "Abort entire simulation (HARD) or just the current time step (SOFT).");
  params.addParam<std::string>(
      "message", "An optional message to be output when the termination condition is triggered");

  MooseEnum errorLevel("INFO WARNING ERROR");
  params.addParam<MooseEnum>(
      "error_level",
      errorLevel,
      "The error level for the message. A level of ERROR will always lead to a hard "
      "termination of the entire simulation.");
  
  return params;
}

Terminator::Terminator(const InputParameters & parameters)
  : GeneralUserObject(parameters),
    _fail_mode(getParam<MooseEnum>("fail_mode").getEnum<FailMode>()),
    _error_level(isParamValid("error_level")
                     ? getParam<MooseEnum>("error_level").getEnum<ErrorLevel>()
                     : ErrorLevel::NONE),
    _vpp_names(getParam<std::vector<VectorPostprocessorName>>("vpp_names")),
    _criterion(getParam<double>("criterion"))
{
  // sanity check the parameters
  if (_error_level == ErrorLevel::ERROR && _fail_mode == FailMode::SOFT)
    paramError("error_level", "Setting the error level to ERROR always causes a hard failure.");
  if (_error_level != ErrorLevel::NONE && !isParamValid("message"))
    paramError("error_level",
               "If this parameter is specified a `message` must be supplied as well.");
  _vpp_num = _vpp_names.size();
  _vpp_old.resize(_vpp_num);
  _vpp_old_holder.resize(_vpp_num);
  _vpp_old_old.resize(_vpp_num);
}

void
Terminator::handleMessage()
{
  if (!isParamValid("message"))
    return;

  auto message = getParam<std::string>("message");
  switch (_error_level)
  {
    case ErrorLevel::INFO:
      mooseInfo(message);
      break;

    case ErrorLevel::WARNING:
      mooseWarning(message);
      break;

    case ErrorLevel::ERROR:
      mooseError(message);
      break;

    default:
      break;
  }
}

void
Terminator::execute()
{
  converged = true;
  for (unsigned int i = 0; i < _vpp_num; ++i) {
    _fe_problem.terminateSolve();
    std::cout << _vpp_names[i] << "\n";
    const VectorPostprocessorValue* value_column = &getVectorPostprocessorValueByName(_vpp_names[i], "flux_aggregate");
    VectorPostprocessorValue old_old_value_column = _vpp_old_old[i];
    if (old_old_value_column.size() == 0)
      old_old_value_column.resize(value_column->size());
    _vpp_old_holder[i] = *value_column;
    for (int j=0; j < value_column->capacity(); j++) {
      std::cout << j << ": NEW " << value_column->at(j) << "\n";
      std::cout << j << ": OLD_OLD " << old_old_value_column[j] << "\n";
      double check_val = std::abs(value_column->at(j) - old_old_value_column[j]);
      if (check_val > _criterion) {
        converged = false;
        break;
      }
    }
  }
  _vpp_old_old = _vpp_old;
  _vpp_old = _vpp_old_holder;

  // request termination of the run or timestep in case the expression evaluates to true
  if (converged)
  {
    if (_fail_mode == FailMode::HARD)
    {
      printf("REACHED HERE WHERE IT SHOULD HAVE\n");
      handleMessage();
      _fe_problem.terminateSolve();
    }
    else
    {
      _console << name() << " is marking the current solve step as failed.\n";
      handleMessage();
      getMooseApp().getExecutioner()->picardSolve().failStep();
    }
  }
}

#endif
```

Note: 
- There's a bunch of stuff left in the code from the old Terminator that doesn't really need to be there any more.
- Convergence is based on comparing the nth non-linear step temperatures and the (n-2)th non-linear step's termperatures.

**Comment by aeslaughter** [comment]
Are you setting the "execute_on" parameter? If not, set this to LINEAR and see if it works.

From a software design perspective, you should not modify the existing object. You should just create a new object that inherits from `GeneralUserObject` or `Terminator`, depending if you need some of the same features. The object you create should go into your application.

**Reply by wrkendrick** [reply]
> Are you setting the "execute_on" parameter? If not, set this to LINEAR and see if it works.

Yep, currently running every non-linear, tried on linear and still getting the same deal. If I set the error_level to WARNING then the warning goes off when it should (each nonlinear step), but the run fails to terminate, suggesting `_fe_problem.terminateSolve()` may not be able to interrupt mid-run?

> From a software design perspective, you should not modify the existing object. You should just create a new object that inherits from GeneralUserObject or Terminator, depending if you need some of the same features. The object you create should go into your application.

You're absolutely correct, I'm being quick and sloppy right now, I intend to clean it all up and write it the proper way once I can confirm it works as intended.

**Reply by aeslaughter** [reply]
Looked at the code a bit and you are right, it will only terminate on timesteps. The check for the termination flag occurs in [Transient::keepGoing](https://github.com/idaholab/moose/blob/1f37a175d95167ece3db4c3281883988ed5369a9/framework/src/executioners/Transient.C#L583), which is called by [Transient::execute](https://github.com/idaholab/moose/blob/1f37a175d95167ece3db4c3281883988ed5369a9/framework/src/executioners/Transient.C#L308).

@fdkong Is there a supported way to terminate the solve during the linear/non-linear iterations?

**Comment by fdkong** [comment]
>I decided to follow in the footsteps of the Terminator.C user object and focus on using the change in boundary temperatures as my criterion for convergence.

It is not right approach to me if you want to have the simulation converged based your rules. 

You should  check the function  `checkNonlinearConvergence`. You could derive from FEProblem, and write your own version of `checkNonlinearConvergence`

**Comment by permcody** [comment, marked as answer]
I don't believe Steady (a single solve) is the right way to approach this problem. With complex problem, you are likely to get a more robust solve if you can break it into "steps". Even if there is no time dependency in your equation set you might choose to ramp up boundary or body forces over a series of solve steps. This is common in solid mechanics and is called "load steps". If you do that, you would use the "Transient" executioner with perhaps a fake or non-dimensional time (used to control your load steps) and your terminator condition would also just naturally work as-is.
